core/lib/page-functions.js

```diff
@@ -513,39 +515,95 @@ function truncate(string, characterLimit) {
+  // TODO: refactor and delete `global.isDevtools`.
+  if (global.isDevtools || global.isLightrider) return true;
+
+  const require = createRequire(import.meta.url);
+
+  try {
+    // Not foolproof, but `lighthouse-logger` is a dependency of lighthouse that should always be resolvable.
+    // `require.resolve` will only throw in atypical/bundled environments.
+    require.resolve('lighthouse-logger');
+    return false;
+  } catch (err) {
+    return true;
+  }
+}
+
 // This is to support bundled lighthouse.
 // esbuild calls every function with a builtin `__name` (since we set keepNames: true),
 // whose purpose is to store the real name of the function so that esbuild can rename it to avoid
 // collisions. Anywhere we inject dynamically generated code at runtime for the browser to process,
 // we must manually include this function (because esbuild only does so once at the top scope of
 // the bundle, which is irrelevant for code executed in the browser).
-const esbuildFunctionNameStubString = 'var __name=(fn)=>fn;';
+// When minified, esbuild will mangle the name of this wrapper function, so we need to determine what it
+// is at runtime in order to recreate it within the page.
+const functionAsString = (()=>{
+  // eslint-disable-next-line no-unused-vars
+  const a = ()=>{};
+}).toString()
+  // When not minified, esbuild annotates the call to this function wrapper with PURE.
+  // We know further that the name of the wrapper function should be `__name`, but let's not
+  // hardcode that. Remove the PURE annotation to simplify the regex.
+  .replace('/* @__PURE__ */', '');
+const functionStringMatch = functionAsString.match(/=\s*([\w_]+)\(/);
+if (!functionStringMatch && isBundledEnvironment()) {
+  throw new Error('Could not determine esbuild function wrapper name');
+}
 
-/** @type {string} */
-const truncateRawString = truncate.toString();
-truncate.toString = () => `function truncate(string, characterLimit) {
+let esbuildFunctionWrapperString = '';
+if (functionStringMatch) {
+  esbuildFunctionWrapperString =
+    // eslint-disable-next-line max-len
+    `var ${functionStringMatch[1]}=(fn, value) => Object.defineProperty(fn, 'name', {value, configurable: true});`;
+}
+
+/**
+ * @param {Function} fn
+ * @return {string}
+ */
+function getRuntimeFunctionName(fn) {
+  const match = fn.toString().match(/function ([\w$]+)/);
+  if (!match) throw new Error(`could not find function name for: ${fn}`);
+  return match[1];
+}
+getRuntimeFunctionName(truncate);
+
+// We setup a number of our page functions to automatically include their dependencies.
+// Because of esbuild bundling, we must refer to the actual (mangled) runtime function name.
+/** @type {Record<string, string>} */
+const names = {
+  truncate: getRuntimeFunctionName(truncate),
+  getNodeLabel: getRuntimeFunctionName(getNodeLabel),
+  getOuterHTMLSnippet: getRuntimeFunctionName(getOuterHTMLSnippet),
+  getNodeDetails: getRuntimeFunctionName(getNodeDetails),
+};
```

Comment:
nit: my preference would be for these to be standalone variables (e.g. `truncateName`). Is there a a a reason these are attached to a record object. It's kind confusing since `names` only contains the name of *some* functions.